app/models/synonym.rb

```diff
@@ -30,38 +30,83 @@
+
+  include ArelHelpers::ArelTable
+
   has_many :names
 
-  # Mightly cronjob to ensure that no synonym records accidentally got deleted.
+  # Nightly cronjob to ensure that no synonym records accidentally got deleted.
   # This actually happened to Fungi itself(!)  Not sure how it happened, and
   # obviously I'd prefer to fix the cause.  But meanwhile, might as well keep
   # the site working...
   def self.make_sure_all_referenced_synonyms_exist
     msgs = []
-    references = Name.connection.select_values(%(
-      SELECT DISTINCT synonym_id FROM names
-      WHERE synonym_id IS NOT NULL
-      ORDER BY synonym_id ASC
-    ))
-    records = Name.connection.select_values(%(
-      SELECT id FROM synonyms ORDER BY id ASC
-    ))
+    # reference_select = arel_select_referenced_synonyms
+    # # puts(reference_select.to_sql)
+    # references = Name.connection.select_values(reference_select.to_sql)
+    names = Name.arel_table
+    references = Name.where(names[:synonym_id].not_eq(nil)).distinct.
+                 order(synonym_id: :asc).pluck(:synonym_id)
+    # puts(references.join(",").to_s)
+
+    # SELECT id FROM synonyms ORDER BY id ASC
+    # record_select = Synonym.select(:id).order(Synonym[:id].asc)
+    # puts(record_select.to_sql)
+    # records = Name.connection.select_values(record_select.to_sql)
+    records = Synonym.all.order(id: :asc).pluck(:id)
+    # puts(records.join(",").to_s)
     unused  = records - references
     missing = references - records
+    # puts(unused.join(",").to_s)
+    # puts(missing.join(",").to_s)
+
     if unused.any?
-      Name.connection.execute(%(
-        DELETE FROM synonyms
-        WHERE id IN (#{unused.map(&:to_s).join(",")})
-      ))
+      delete_manager = arel_delete_unused_synonyms(unused)
+      # puts(delete_manager.to_sql)
+      Name.connection.delete(delete_manager.to_sql)
+
       msgs << "Deleting #{unused.count} unused synonyms: #{unused.inspect}"
     end
     if missing.any?
-      Name.connection.execute(%(
-        INSERT INTO synonyms (id) VALUES
-        #{missing.map { |id| "(#{id})" }.join(",")}
-      ))
+      insert_manager = arel_insert_missing_synonyms(missing)
+      # puts(insert_manager.to_sql)
+      Name.connection.execute(insert_manager.to_sql)
       msgs << "Restoring #{missing.count} missing synonyms: #{missing.inspect}"
     end
     msgs
   end
+
+  # SELECT DISTINCT synonym_id FROM names
+  # WHERE synonym_id IS NOT NULL
+  # ORDER BY synonym_id ASC
+  # private_class_method def self.arel_select_referenced_synonyms
+  #   names = Name.arel_table
+
+  #   Name.select(:synonym_id).distinct.
+  #     where(names[:synonym_id].not_eq(nil)).
+  #     order(names[:synonym_id].asc)
+  # end
+
+  # DELETE FROM synonyms
+  # WHERE id IN (#{unused.map(&:to_s).join(",")})
+  private_class_method def self.arel_delete_unused_synonyms(unused)
+    syn = Synonym.arel_table
+
+    Arel::DeleteManager.new.
+      from(syn).
+      where(syn[:id].in(unused))
+  end
+
```

Comment:
I agree with you about using AR rather than Arel provided that AR doesn't slow things down too much.
I would use `delete_all` even if it leads to inconsistency. 
What's the advantage of the `mass_insert` gem, given that you can do it in Arel?
IMO delete vs destroy is outside the scope of this PR: `delete_all` is a fine replacement for the SQL Any problems caused by avoiding callbacks were already there, and we should worry about them another day.